File: 프로그래머스/3/42898. 등굣길/등굣길.py
# m x n 크기의 격자모양
# puddles : 물이 잠긴 지역의 좌표 담은 2차원 배열
# 오른쪽과 아래쪽으로만 움직여 집까지 갈 수 있는 경로의 개수를 1000000007로 나눈 나머지
# 점화식 :     grid[i][j] = grid[i-1][j] + grid[i][j-1]
# grid[1][1] = 1
def solution(m, n, puddles):
    grid = [[0] * (m+1) for _ in range(n+1)]
    grid[1][1] = 1
    
    for r, c in puddles:
        grid[c][r] = -1
    
    for row in range(1, n + 1):
        for col in range(1, m + 1):
            if grid[row][col] == -1:
                continue
            if grid[row-1][col] != -1:
                grid[row][col] += grid[row-1][col] 
            if grid[row][col-1] != -1:
                grid[row][col] += grid[row][col-1]
    
    return grid[n][m] % 1000000007
    

File: 프로그래머스/3/42898. 등굣길/등굣길.java
class Solution {
    // 집에서 학교까지 가는 길은 mxn 크기의 격자모양
    // puddles : 물이 잠긴 지경의 좌표
    // return : 오른쪽과 아래쪽으로만 움직여 집에서 학교까지 갈 수 있는 최단경로의 개수%1000000007
    private static int[][] memo;
    public int dp(int row, int col) {
        if(row < 1 || col < 1) return 0;
        
        if(memo[row][col] == -1) return 0;
        if(memo[row][col] != 0) return memo[row][col];
        return memo[row][col] = (dp(row -1, col) + dp(row, col-1))%1000000007;
    }
    
    public int solution(int m, int n, int[][] puddles) {
        memo = new int[n+1][m+1];
        memo[1][1] = 1;
        for(int[] p : puddles) memo[p[1]][p[0]] = -1;
        
        dp(n, m);
        return memo[n][m];
    }
}